webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/campaigns/JsonCampaignSerializer.scala

```diff
@@ -0,0 +1,168 @@
+*/
+
+package com.normation.rudder.campaigns
+
+import com.normation.errors.IOResult
+import com.normation.errors.Inconsistency
+import com.normation.utils.DateFormaterService
+import org.joda.time.DateTime
+import org.joda.time.format.DateTimeFormat
+import zio.json.DeriveJsonDecoder
+import zio.json.DeriveJsonEncoder
+import zio.json.JsonDecoder
+import zio.json.JsonEncoder
+import zio.json.ast.Json
+import zio.syntax._
+
+class CampaignSerializer {
+
+  private[this] var tranlaters : List[JSONTranslateCampaign] = Nil
+
+  def getJson(campaign: Campaign) = tranlaters.map(_.getRawJson()).fold(Jsonbase) { case (a, b) => b orElse a }(campaign)
+
+
+  val Jsonbase : PartialFunction[Campaign, IOResult[Json]] = {
+    case c : Campaign => Inconsistency(s"No translater for campaign ${c.info.id.value}").fail
+  }
+
+  val base : PartialFunction[Campaign, IOResult[String]] = {
+    case c : Campaign => Inconsistency(s"No translater for campaign ${c.info.id.value}").fail
+  }
+
+  val readBase : PartialFunction[String, IOResult[Campaign]]= {
+    case c : String => Inconsistency(s"could not translate into campaign").fail
+  }
+
+  def addJsonTranslater(c : JSONTranslateCampaign ) = tranlaters = c :: tranlaters
+
+  def serialize(campaign: Campaign) : IOResult[String] = tranlaters.map(_.handle()).fold(base) { case (a, b) => b orElse a }(campaign)
+  def parse(string : String) : IOResult[Campaign] = tranlaters.map(_.read()).fold(readBase) { case (a, b) => b orElse a }(string)
+
+}
+
+trait JSONTranslateCampaign{
+
+  def getRawJson(): PartialFunction[Campaign, IOResult[Json]]
+
+  def handle(): PartialFunction[Campaign, IOResult[String]]
+
+  def read(): PartialFunction[String, IOResult[Campaign]]
+
+}
+
+object JSONTranslateCampaign {
+  implicit val idEncoder : JsonEncoder[CampaignId]= JsonEncoder[String].contramap(_.value)
+  implicit val dateTime : JsonEncoder[DateTime] = JsonEncoder[String].contramap(DateFormaterService.serialize)
+  implicit val dayOfWeek : JsonEncoder[DayOfWeek] = JsonEncoder[Int].contramap(_.value)
+  implicit val monthlySchedulePosition : JsonEncoder[MonthlySchedulePosition] =  JsonEncoder[Int].contramap(
+    s =>
+      s match {
+        case First => 1
+        case Second => 2
+        case Third => 3
+        case Last => -1
+        case SecondLast => -2
+      }
+  )
+  implicit val statusInfoEncoder   : JsonEncoder[CampaignStatus] = DeriveJsonEncoder.gen
+  implicit val scheduleEncoder : JsonEncoder[CampaignSchedule]= DeriveJsonEncoder.gen
+  implicit val campaignInfoEncoder : JsonEncoder[CampaignInfo]= DeriveJsonEncoder.gen
+
+  implicit val idDecoder : JsonDecoder[CampaignId] = JsonDecoder[String].map(s => CampaignId(s))
+
+  val iso8601 = DateTimeFormat.forPattern("yyyy-MM-dd'T'HH:mm:ssZZ")
+  implicit val decodeIsoDate: JsonDecoder[DateTime] = JsonDecoder[String].mapOrFail(s =>
+    try {
+      Right(iso8601.parseDateTime(s))
+    } catch {
+      case e: Exception => Left(e.getMessage)
+    }
+  )
```

Comment:
actually, that devoder should be in dateformatter service (but that's for an other day, because zio json is not at that level of pom.xml IIRC)